openshift/assisted-service
# MGMT-12198: Validate Install Config Overrides

[MGMT-12198](https://issues.redhat.com/browse/MGMT-12198)
Previously if invalid json fields were provided
to the install config overrides, the service
would just ignore them.

This adds a validation so that any additional field added to the install config override will generate an error response for the whole request.

New error message when an invalid field is specified:
```
{"code":"400","href":"","id":400,"kind":"Error","reason":"json: unknown field \"foo\""}
```



## List all the issues related to this PR


## What environments does this code impact?


## How was this code tested?




#### Manual test
1. Created a cluster
    ````
    $ curl ${ASSISTED_SERVICE_URL}/api/assisted-install/v2/clusters -X POST -H "Content-Type: application/json" --data '{"name": "test-cluster", "openshift_version":"4.9","base_dns_domain":"redhat.com","pull_secret":"${SECRET}"}'
    ````
2. Patched with a valid install config override
    ````
    $ curl ${ASSISTED_SERVICE_URL}/api/assisted-install/v2/clusters/${CLUSTER_ID}/install-config  -X PATCH -H 'accept: application/json'  -H 'Content-Type: application/json' -d '"{\"controlPlane\":{\"hyperthreading\":\"Enabled\"}}"'
    ````
3. Get the cluster's install config override to make sure it applied
    ````
    $ curl ${ASSISTED_SERVICE_URL}/api/assisted-install/v2/clusters/${CLUSTER_ID} | jq '.install_config_overrides'
    "{\"controlPlane\":{\"hyperthreading\":\"Enabled\"}}"
    ````
4. Patched again with an invalid install config override 
    ````
    $ curl ${ASSISTED_SERVICE_URL}/api/assisted-install/v2/clusters/${CLUSTER_ID}/install-config  -X PATCH -H 'accept: application/json'  -H 'Content-Type: application/json' -d '"{\"controlPlane\":{\"hyperthreading\":\"Disabled\"}, \"foo\":\"bar\"}"'
    {"code":"400","href":"","id":400,"kind":"Error","reason":"json: unknown field \"foo\""}
    ````
5. Get the cluster's install config override to make sure it didn't change
    ````
    $ curl ${ASSISTED_SERVICE_URL}/api/assisted-install/v2/clusters/${CLUSTER_ID} | jq '.install_config_overrides'
    "{\"controlPlane\":{\"hyperthreading\":\"Enabled\"}}"
    ````
## Checklist


## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

/cc @carbonin @filanov @eranco74
